moped-editor/src/views/projects/projectView/ToolbarPlugin.js

```diff
@@ -0,0 +1,287 @@
+    label: "Bulleted list",
+  },
+  {
+    id: richTextAction.link,
+    icon: <Link />,
+    label: "Link",
+  },
+  { id: richTextAction.divider },
+  {
+    id: richTextAction.undo,
+    icon: <Undo />,
+    label: "Undo",
+  },
+  {
+    id: richTextAction.redo,
+    icon: <Redo />,
+    label: "Redo",
+  },
+  {
+    id: richTextAction.clear,
+    icon: <FormatClear />,
+    label: "Clear formatting",
+  },
+];
+
+const ToolbarPlugin = ({ noteAddSuccess, classes }) => {
+  const [editor] = useLexicalComposerContext();
+
+  const [disableMap, setDisableMap] = useState({
+    [richTextAction.undo]: true,
+    [richTextAction.redo]: true
+  });
+  const [selectionMap, setSelectionMap] = useState({});
+
+  // Iterate through selected node and parent nodes and returns any applicable list type
+  const checkListType = (selection, listType) => {
+    let hasListType = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((node) => {
+      const nodeParents = node.getParents();
+      nodeParents.forEach((parent) => {
+        if (parent.__listType === listType) {
+          hasListType = true;
+        }
+      })
+    })
+    return hasListType;
+  };
+
+  // Checks for selected list formatting, removes if already applied or adds if not applied
+  const handleListUpdate = useCallback((command, listType) => {
+    const isListApplied = selectionMap[listType];
+    isListApplied ?
+      editor.dispatchCommand(REMOVE_LIST_COMMAND) :
+      editor.dispatchCommand(command);
+  }, [editor, selectionMap]);
+
+  const checkLink = (selection) => {
+    let isLink = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((selectedNode) => {
+      const selectedNodeParent = selectedNode.getParent()
+      if ($isLinkNode(selectedNodeParent)) {
+        isLink = true
+      }
+    })
+    return isLink;
+  };
+
+  const handleLinkUpdate = useCallback(() => {
+    const isLinkApplied = selectionMap["link"];
+    isLinkApplied ?
+      editor.dispatchCommand(TOGGLE_LINK_COMMAND, null)
+      :
+      navigator.clipboard.readText().then(val => {
+        editor.dispatchCommand(TOGGLE_LINK_COMMAND, {
+          url: val, target: "_blank"
+        })
+      }).catch(err => {
+        console.error('Failed to read clipboard contents: ', err);
+      })
+  }, [editor, selectionMap]);
+
+  const updateToolbar = useCallback(() => {
+    const selection = $getSelection();
+    if ($isRangeSelection(selection)) {
+      const newSelectionMap = {
+        [richTextAction.bold]: selection.hasFormat("bold"),
+        [richTextAction.italics]: selection.hasFormat("italic"),
+        [richTextAction.underline]: selection.hasFormat("underline"),
+        [richTextAction.number]: checkListType(selection, "number"),
+        [richTextAction.bullet]: checkListType(selection, "bullet"),
+        [richTextAction.link]: checkLink(selection),
+      }
+      setSelectionMap(newSelectionMap)
+    }
+  }, [setSelectionMap])
```

Comment:
i'm surprised this isn't throwing an error about `checkListType`, `checkLink`, and `richTextAction` not being in the dependency array. i have doubts about the right way to set this up, but maybe an empty dependency array is actually what's desired here?

Reply:
you're right, not sure why it wasn't throwing an error about those missing dependencies, but feeding in an empty dependency array does seem like the way to go! thanks

Reply:
yes, i think that now that the `checkListType` and `checkLink` functions are moved out of the component, then an empty dependency array makes sense. Side note that I recently found that the new React docs don't mention state setter functions have stable references between renders but [the old ones do](https://legacy.reactjs.org/docs/hooks-reference.html#usestate).

Reply:
Mike, your comment sent me digging through the new docs out of curiosity, and there is a comment about "stable identity" in these [setState caveats](https://react.dev/reference/react/useState#setstate-caveats) 😅